File: courseMaterial/course_ROSActions/code_section_3_simple_actions_activity/my_robot_tutorials/scripts/move_robot_client.py
#!/usr/bin/env python

import rospy
import actionlib
from std_msgs.msg import Empty
from my_robot_msgs.msg import MoveRobotAction
from my_robot_msgs.msg import MoveRobotGoal

class MoveRobotClient:

    def __init__(self):
        self._ac = actionlib.SimpleActionClient('/move_robot',
            MoveRobotAction)
        self._ac.wait_for_server()
        rospy.loginfo("Server is up, let's send a goal!")

        self.sub = rospy.Subscriber('/cancel_move', Empty, 
            self.cancel_move_callback)

    def cancel_move_callback(self, req):
        rospy.loginfo("Received message to cancel goal")
        self._ac.cancel_goal()

    def send_goal(self):
        goal = MoveRobotGoal()
        goal.position = 99
        goal.velocity = 1
        self._ac.send_goal(goal, done_cb=self.done_callback,
            feedback_cb=self.feedback_callback)
        rospy.loginfo("Goal has been sent")

    def done_callback(self, status, result):
        rospy.loginfo("Status: " + str(status))
        rospy.loginfo("Result: " + str(result))

    def feedback_callback(self, feedback):
        rospy.loginfo(feedback)

if __name__ == '__main__':
    rospy.init_node('move_robot_client')

    client = MoveRobotClient()
    client.send_goal()

    rospy.spin()

File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/src/move_robot_client.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>

#include <std_msgs/Empty.h>

#include <my_robot_msgs/MoveRobotAction.h>
#include <my_robot_msgs/MoveRobotGoal.h>
#include <my_robot_msgs/MoveRobotResult.h>
#include <my_robot_msgs/MoveRobotFeedback.h>

class MoveRobotClient {
	
	protected:
	ros::NodeHandle _nh;
	actionlib::SimpleActionClient<my_robot_msgs::MoveRobotAction> _ac;
	ros::Subscriber _cancel_subscriber;

	public:
	MoveRobotClient():
		_ac("/move_robot", true)
	{
		ROS_INFO("Wait for the action server to start...");
		_ac.waitForServer();
		ROS_INFO("Server is now up");

		_cancel_subscriber = _nh.subscribe("/cancel_move", 10,
			&MoveRobotClient::cancelMoveCallback, this);
	}

	void cancelMoveCallback(const std_msgs::Empty &msg)
	{
		ROS_INFO("Received message to cancel goal");
		_ac.cancelGoal();
	}

	void sendGoal() {
		my_robot_msgs::MoveRobotGoal goal;
		goal.position = 99;
		goal.velocity = 4;
		_ac.sendGoal(goal,
			boost::bind(&MoveRobotClient::doneCb, this, _1, _2),
			boost::bind(&MoveRobotClient::activeCb, this),
			boost::bind(&MoveRobotClient::feedbackCb, this, _1));
		ROS_INFO("Goal has been sent");
	}

	void doneCb(const actionlib::SimpleClientGoalState& state,
		const my_robot_msgs::MoveRobotResultConstPtr &result)
	{
		ROS_INFO("Finished in state: %s",
			state.toString().c_str());
		ROS_INFO("Move robot result: %d", (int)result->position);
	}

	void activeCb() 
	{
		ROS_INFO("Goal just went active");
	}	

	void feedbackCb(const my_robot_msgs::MoveRobotFeedbackConstPtr &feedback)
	{
		ROS_INFO("Feedback received. Current position: %d", 
			(int)feedback->current_position);
	}
};

int main (int argc, char **argv) 
{
	ros::init(argc, argv, "move_robot_client");
	MoveRobotClient client;
	client.sendGoal();
	ros::spin();
}
